### ISSUE TITLE
support for "faces"

### ISSUE THREAD

**Issue by rabernat** [xgcm/xgcm#5]
New MITgcm runs have multiple cube faces. We need to figure out how to deal with that.

**Comment by lesommer** [comment]
Hi Ryan, could you describe a bit this notion of cube faces ?

**Comment by rabernat** [comment]
The idea was developed by Gael Forget in in GCMFaces MATLAB package:
http://mitgcm.org/viewvc/*checkout*/MITgcm/MITgcm_contrib/gael/matlab_class/gcmfaces.pdf
![image](https://cloud.githubusercontent.com/assets/1197350/15673995/51b3a44e-2708-11e6-981c-0ad17b9992b2.png)

The latest MITgcm global simulations use rather complex grid topologies. The "faces" concept involves splitting the domain into several faces (i.e. large tiles) that are locally "regular" and operating individually on those.

See also http://www.geosci-model-dev.net/8/3071/2015/gmd-8-3071-2015.pdf

**Comment by lesommer** [comment]
and I guess each "face" corresponds to a particular netCDF file, right ?

**Comment by rabernat** [comment]
>I guess each "face" corresponds to a particular netCDF file, right ?

No, not at all.

These runs generally do not use netCDF output. They use MITgcm's MDS binary file format. [This mess](https://github.com/xgcm/xgcm/blob/master/xgcm/mdsxray.py) is required to get the MDS files into xarray, and that code doesn't even support the LLC/CS grids. With faces, the logical layout within the binary output files is very different from the physical layout.

**Comment by lesommer** [comment]
Now I understand why you opened a github issue here. 

Do you think this concept of "face" would be useful for describing hierarchies of structured grids with different resolution ?

**Comment by rabernat** [comment]
I'm not sure about nested grids...would have to think a bit more.

The way I imagine it now, the `face` would be an additional dimension. As in
```
<xarray.DataArray (time: 12, face: 5, zc: 60, yc: 512, xc: 512)>
```
Then plotting / integration methods can iterate over the faces and do what they have to do.

If your nested grid can be massaged into such a format, then it would work. The problem is that the nested sub-grids are unlikely to all have the same shape.

**Comment by lesommer** [comment]
Indeed, sub-grids would not have the same shape. 

How do you think the concept of `grid` used in oocgcm could be used with MITgcm cube faces ? You would probably have to define a grid object for each face, right ?

**Comment by rabernat** [comment]
If `face` can be represented as an additional dimension (as in my example above), then a separate grid object is probably _not_ required for each face. Xarray will automatically align any arithmetic operations along faces. Some special logic may be required for the face boundaries though.

If `face` _cannot_ be represented as an additional dimension (as with the nested grid), then I guess separate grid objects would be necessary. But then you couldn't even store the data using xarray, since it doesn't support such complex hierarchies. So you would have much bigger problems.

**Comment by rabernat** [comment]
This is fixed by xgcm/xmitgcm#4
